MediaQuery picks up view data changes

justinmc opened the pull request:
This fixes a bug in MediaQuery's update logic and removes some now-unnecessary workaround code.

Fixes https://github.com/flutter/flutter/issues/165519

Discovered in https://github.com/flutter/flutter/pull/165354#discussion_r1999849170

justinmc committed:
MediaQuery updates when view data changes

justinmc committed:
Formatting

justinmc committed:
Test the change

Renzo-Olivares commented on packages/flutter/lib/src/widgets/media_query.dart:
```diff
@@ -1882,7 +1882,9 @@ class _MediaQueryFromViewState extends State<_MediaQueryFromView> with WidgetsBi
       _updateParentData();
     }
-    if (_data == null || oldWidget.view != widget.view) {
+    if (_data == null ||
```
A pattern I notice in this file is `didChange*()` which updates the `MediaQueryData`. These notifications are sent from the embedder. I wonder if that's the pattern we should be following here, what do you think?

justinmc replied:
Ah you're saying that I could add a WidgetsBindingObserver.didChangeViewData and call that instead of didChangeMetrics? In this case I only need the method for tests, so I worry about bloating the  WidgetsBindingObserver API just for my that. And anyway I'd still have to use my hacky _updateMediaQueryFromView method in the tests.

I looked into seeing if there's a better way to send these kinds of messages from the embedder to trigger didChangeMetrics instead of directly calling it on the  WidgetsBindingObserver instance, but I found other tests were doing the same thing as me, e.g. https://github.com/flutter/flutter/blob/2615ab6c1932e44e6802aaba8dc715b387ff155f/packages/flutter/test/widgets/editable_text_test.dart#L15928

It seems to me like this should be checking if the data has changed anyway.

justinmc replied:
After talking to @Renzo-Olivares offline, I'm now trying to directly trigger _updateData when supportsShowingContextMenu changes. Inspired by https://github.com/flutter/flutter/pull/166836.

justinmc replied:
@Renzo-Olivares I've set this up so that I can trigger the change through the PlatformDispatcher, similar to changes to the theme brightness etc. Let me know what you think.

justinmc committed:
Merge branch 'master' into media-query-view-data-update

justinmc committed:
Try directly notifying that supportsShowingContextMenu changed

justinmc committed:
analyzer

justinmc committed:
Merge branch 'master' into media-query-view-data-update

justinmc committed:
Manually call update in test and it works

justinmc committed:
Cleanup

justinmc committed:
no binding needed

justinmc committed:
Fix mediaquery test

justinmc committed:
Fix docs

justinmc committed:
Web support

justinmc committed:
Changed not Change

justinmc committed:
Missing implementation

Renzo-Olivares commented on packages/flutter/test/widgets/media_query_test.dart:
```diff
@@ -262,6 +262,57 @@ void main() {
     },
   );
 
+  testWidgets('MediaQueryData.fromView picks up new view data when rebuilding', (
```
nit: should we rename this test to mention support showing system context menu.

Renzo-Olivares commented on packages/flutter/test/widgets/media_query_test.dart:
```diff
@@ -262,6 +262,57 @@ void main() {
+
+    tester.platformDispatcher.supportsShowingSystemContextMenu = true;
+    tester.binding.handleSupportsShowingSystemContextMenuChanged();
+    addTearDown(() {
```
Can these be added to the `addTearDown` at the top of this test?

justinmc replied:
Yes good call.

Renzo-Olivares commented on packages/flutter/test/widgets/media_query_test.dart:
```diff
@@ -262,6 +262,57 @@ void main() {
+    expect(datas1.first.supportsShowingSystemContextMenu, isFalse);
+
+    tester.platformDispatcher.supportsShowingSystemContextMenu = true;
+    tester.binding.handleSupportsShowingSystemContextMenuChanged();
```
Does this have to be explicitly called?

Renzo-Olivares replied:
Yeah it looks like your current implementation is in line with the pattern we use for other APIs in `platformDispatcher`.

justinmc replied:
That test confuses me because it sets the brightness to `light`, then expects that it's still `dark`...

But this is my understanding of the situation as I've been working on this PR. Each time you test a file with `flutter test`, a single top-level View gets created. When you run `pumpWidgets`, those widgets get put into this View, and the View has [its own top-level MediaQuery](https://github.com/flutter/flutter/blob/6b6f49e74923fa872e246895b548496403867d0c/packages/flutter/lib/src/widgets/view.dart#L275) widget. So even though you might call `pumpWidgets` with a whole new MaterialApp, you still have the same old top level MediaQuery widget.

When I change the view data with something like:

```dart
tester.platformDispatcher.supportsShowingSystemContextMenu = true;
```

That MediaQuery widget does not automatically update. It still has its old view data until you call [_updateData](https://github.com/flutter/flutter/blob/e841350bfc86a3d3e0377cddb8974594e752fc15/packages/flutter/lib/src/widgets/media_query.dart#L1896) one way or another. In my case, I've set up handleSupportsShowingSystemContextMenuChanged to call _updateData.

Note that this situation doesn't matter when running a single test. If you set tester.platformDispatcher.supportsShowingSystemContextMenu in the initial `setup` before that top-level MedaiQuery.fromView gets instantiated, then it will pick up your value when it does get created.

justinmc committed:
Analyzer

justinmc committed:
Merge branch 'master' into media-query-view-data-update

justinmc committed:
Test naming

Renzo-Olivares commented on engine/src/flutter/lib/ui/platform_dispatcher.dart:
```diff
@@ -1142,6 +1142,17 @@ class PlatformDispatcher {
+  Zone _onSupportsShowingSystemContextMenuChangedZone = Zone.root;
+  set onSupportsShowingSystemContextMenuChanged(VoidCallback? callback) {
+    _onSupportsShowingSystemContextMenuChanged = callback;
+    _onSupportsShowingSystemContextMenuChangedZone = Zone.current;
```
Why was the analyzer complaining about the `Zone` related code removed from the `web_ui/lib/src/engine/platform_dispatcher.dart`?

justinmc replied:
It complained that it was unused. I had just copied it from the similar code for textScaler or something.

justinmc committed:
Merge branch 'master' into media-query-view-data-update

justinmc committed:
Merge branch 'master' into media-query-view-data-update

justinmc committed:
Attempt at using wrapWithView instead of all the 'changed' infrastructure.

Doing this blind as I was struggling to run the test locally with a
local engine, but actually after this change I won't need a custom
engine.

The next commit will double check the tests and make any necessary
changes.

justinmc committed:
Fix all tests for new wrapWithView approach.

Needed to pump a second View at the end of the test to clear the value
and avoid breaking other tests.

justinmc committed:
Remove unused handleSupportsShowingSystemContextMenuChanged

justinmc committed:
Unnecessary media query test

justinmc commented:
@Renzo-Olivares Ready for re-review. Thanks for the suggestion here. Your approach worked, but I had to also add an extra pumpWidget in the teardown in order to reset the value for subsequent tests.

Renzo-Olivares commented on packages/flutter/test/cupertino/text_field_test.dart:
```diff
@@ -8839,18 +8839,29 @@ void main() {
+          // Pump one more root view to replace it with one that reads the
+          // updated reset value of supportsShowingSystemContextMenu for
+          // subsequent tests..
+          await tester.pumpWidget(
```
I think we can use `tester.view.reset()` instead of pumping another root view. Found that here https://github.com/flutter/flutter/blob/5da19ccfbc17643fdd7e00615a369070d43e2055/packages/flutter/test/widgets/media_query_test.dart#L407

justinmc committed:
Analyzer

justinmc committed:
tester.view.reset instead of manually pumping a new view

Renzo-Olivares reviewed (approved):
LGTM

justinmc commented:
Thanks @Renzo-Olivares for all the help making this PR clean!
